modules/tinymce/src/plugins/link/main/ts/core/NormalizeLink.ts

```diff
@@ -0,0 +1,54 @@
+import DomTreeWalker from 'tinymce/core/api/dom/TreeWalker';
+import Editor from 'tinymce/core/api/Editor';
+
+const isBR = (elm: Node): elm is HTMLAnchorElement => elm && elm.nodeName.toLowerCase() === 'br';
+
+const isAnchor = (elm: Node): elm is HTMLAnchorElement => elm && elm.nodeName.toLowerCase() === 'a';
+
+const isShortEnded = (editor: Editor, elm: Node) => elm && editor.schema.getShortEndedElements().hasOwnProperty(elm.nodeName);
+
+const isValidNode = (editor: Editor, node: Node) => !isShortEnded(editor, node) && isAnchor(node);
+
+const getClosestLink = (editor: Editor, node: Node) => editor.dom.getParent(node, 'a[href]');
+
+const hasSameParent = (editor: Editor, elm: Node, nextNode: Node): elm is HTMLAnchorElement =>
+  editor.dom.getParent(elm, 'p') === editor.dom.getParent(nextNode, 'p');
+
+const getSelectedLink = (editor: Editor, selectedElm?: Element): HTMLAnchorElement | null => {
+  const rng = editor.selection.getRng();
+  const startRng = rng.startContainer;
+  const endRng = rng.endContainer;
+
+  const endOffsetChild = endRng.childNodes[rng.endOffset];
+  const startOffsetChild = startRng.childNodes[rng.startOffset];
```

Comment:
There is a utility function you could use for this to resolve the indexes and it does the checking as well.
It's exposed at public api at:
tinymce.dom.RangeUtils.getNode(container, offset)

This is the function that gets called:
https://github.com/tinymce/tinymce/blob/develop/modules/tinymce/src/core/main/ts/selection/RangeNodes.ts#L19

Reply:
nice utility, updated